File: python/hello_world.py
"""
This script is a simple interactive program that asks for a user's name and date of birth,
then calculates their age and determines their zodiac sign!
"""

from datetime import datetime

# Prompt the user to enter their name and store it in the 'name' variable
name = input("What is your name? ")

# Ask for date of birth in yyyyMMdd format
dob_input = input("What is your date of birth? (Enter in yyyyMMdd format) ")

# Parse the date of birth
try:
    dob = datetime.strptime(dob_input, "%Y%m%d")
    today = datetime.now()
    
    # Calculate age
    age = today.year - dob.year - ((today.month, today.day) < (dob.month, dob.day))
    month = dob.month
    day = dob.day
except ValueError:
    print("Invalid date format. Please use yyyyMMdd format.")
    exit()

# Define zodiac signs with accurate date ranges
zodiac_dates = [
    ("Capricorn", (12, 22), (1, 19)),
    ("Aquarius", (1, 20), (2, 18)),
    ("Pisces", (2, 19), (3, 20)),
    ("Aries", (3, 21), (4, 19)),
    ("Taurus", (4, 20), (5, 20)),
    ("Gemini", (5, 21), (6, 20)),
    ("Cancer", (6, 21), (7, 22)),
    ("Leo", (7, 23), (8, 22)),
    ("Virgo", (8, 23), (9, 22)),
    ("Libra", (9, 23), (10, 22)),
    ("Scorpio", (10, 23), (11, 21)),
    ("Sagittarius", (11, 22), (12, 21)),
]

# Function to check if a date falls within a zodiac range
def get_zodiac_sign(month, day):
    for sign, (start_month, start_day), (end_month, end_day) in zodiac_dates:
        if start_month == end_month:
            if month == start_month and start_day <= day <= end_day:
                return sign
        else:  # Sign spans two months
            if (month == start_month and day >= start_day) or (month == end_month and day <= end_day):
                return sign
    return None

# Validate input and determine zodiac sign
if 1 <= month <= 12 and 1 <= day <= 31:
    sign = get_zodiac_sign(month, day)
    if sign:
        print(f"Hello, {name}! You are {age} years old. Your zodiac sign is {sign}!")
    else:
        print(f"Hello, {name}! You are {age} years old. That's not a valid date.")
else:
    print(f"Hello, {name}! You are {age} years old. Please enter a valid date.")

File: c/src/hello_world.c
#include <stdio.h>
#include <string.h>
#include <time.h>

typedef struct {
    char sign[20];
    int start_month;
    int start_day;
    int end_month;
    int end_day;
} ZodiacSign;

const ZodiacSign zodiac_dates[] = {
    {"Capricorn", 12, 22, 1, 19},
    {"Aquarius", 1, 20, 2, 18},
    {"Pisces", 2, 19, 3, 20},
    {"Aries", 3, 21, 4, 19},
    {"Taurus", 4, 20, 5, 20},
    {"Gemini", 5, 21, 6, 20},
    {"Cancer", 6, 21, 7, 22},
    {"Leo", 7, 23, 8, 22},
    {"Virgo", 8, 23, 9, 22},
    {"Libra", 9, 23, 10, 22},
    {"Scorpio", 10, 23, 11, 21},
    {"Sagittarius", 11, 22, 12, 21}
};

char* get_zodiac_sign(int month, int day) {
    for (int i = 0; i < 12; i++) {
        if (zodiac_dates[i].start_month == zodiac_dates[i].end_month) {
            if (month == zodiac_dates[i].start_month && 
                day >= zodiac_dates[i].start_day && 
                day <= zodiac_dates[i].end_day) {
                return (char*)zodiac_dates[i].sign;
            }
        } else {
            if ((month == zodiac_dates[i].start_month && day >= zodiac_dates[i].start_day) ||
                (month == zodiac_dates[i].end_month && day <= zodiac_dates[i].end_day)) {
                return (char*)zodiac_dates[i].sign;
            }
        }
    }
    return NULL;
}

int main() {
    char name[100];
    char dob_input[9];
    int year, month, day;
    int age;
    
    printf("What is your name? ");
    fgets(name, sizeof(name), stdin);
    name[strcspn(name, "\n")] = '\0';  // Remove newline
    
    printf("What is your date of birth? (Enter in yyyyMMdd format) ");
    scanf("%8s", dob_input);
    
    // Parse the date of birth
    if (sscanf(dob_input, "%4d%2d%2d", &year, &month, &day) != 3 ||
        month < 1 || month > 12 || day < 1 || day > 31) {
        printf("Invalid date format. Please use yyyyMMdd format.\n");
        return 1;
    }
    
    // Calculate age
    time_t now = time(NULL);
    struct tm *today = localtime(&now);
    
    age = today->tm_year + 1900 - year;
    if ((today->tm_mon + 1 < month) || 
        (today->tm_mon + 1 == month && today->tm_mday < day)) {
        age--;
    }
    
    // Determine zodiac sign
    char *sign = get_zodiac_sign(month, day);
    
    if (sign != NULL) {
        printf("Hello, %s! You are %d years old. Your zodiac sign is %s!\n", name, age, sign);
    } else {
        printf("Hello, %s! You are %d years old. That's not a valid date.\n", name, age);
    }
    
    return 0;
}
